modules/light-clients/06-solomachine/types/update.go

```diff
@@ -84,6 +89,27 @@ func (cs ClientState) VerifyClientMessage(cdc codec.BinaryCodec, clientMsg expor
 	return nil
 }
 
+// CheckForMisbehaviour returns true for type Misbehaviour (passed VerifyClientMessage check), otherwise returns false
+func (cs ClientState) CheckForMisbehaviour(
+	_ sdk.Context, _ codec.BinaryCodec, _ sdk.KVStore,
+	msg exported.Header, // TODO: Update to exported.ClientMessage
+) bool {
+	if _, ok := msg.(*Misbehaviour); ok {
+		return true
+	}
+
+	return false
+}
+
+// UpdateStateOnMisbehaviour updates state upon misbehaviour. This method should only be called on misbehaviour
+// as it does not perform any misbehaviour checks.
+func (cs ClientState) UpdateStateOnMisbehaviour(
+	_ sdk.Context, _ codec.BinaryCodec, _ sdk.KVStore, // prematurely include args for self storage of consensus state
+) (*ClientState, exported.ConsensusState, error) {
```

Comment:
I'm starting to wonder if we should not return an error? I feel like the code should panic as any errors detected should be returned earlier

Reply:
I agree. I was expecting `VerifyClientMessage` to return an error, `CheckForMisbehaviour` to return a bool and both `UpdateState` and `UpdateStateOnMisbehaviour` to be void.

Reply:
If we just return a bool from `CheckForMisbehavior` we miss out on the ability to return an error message specific to the misbehavior.

Reply:
> we miss out on the ability to return an error message specific to the misbehavior

Can you give an instance of this?